src/main/python/pgshovel/replication/streams/kafka.py

```diff
@@ -40,12 +54,44 @@ def consume(self, state):
             consumer.seek(delta, 1)
             assert consumer.offsets[0] == offset
 
-        for offset, message in consumer:
-            decoded = self.codec.decode(message.value)
-            state = validate_state(state, offset, decoded)
+        for offset, message in self.__primed(consumer, state):
+            state = validate_state(state, offset, message)
             # XXX: This is necessary because of a bug in protocol buffer oneof.
             state = type(state).FromString(state.SerializeToString())
-            yield state, offset, decoded
+            yield state, offset, message
+
+    def __primed(self, consumer, state, attempts=5):
+        decoded = imap(
+            lambda (offset, msg): tuple((offset, self.codec.decode(msg.value))),
+            consumer
+        )
+
+        offset, message = next(decoded)
+
+        try:
+            validate_state(state, offset, message)
+        except InvalidEventError as e:
+            # See if we can drop the first few messages in an attempt to find
+            # the type we're looking for
+            while attempts > 0:
```

Comment:
When a `target` finishes bootstrapping its data with the `loader`, it moves over to the `stream` to start consuming mutations.  Internal to the `KafkaStream` it starts consuming from its kafka `consumer`, validating each message against the `state`, namely the `State.stream_state` state.  When the `stream` *first* starts up, there is no `stream_state`, and as such the transaction validator state machine [validates that the first operation is a `BeginOperation`](https://github.com/disqus/pgshovel/blob/replication/src/main/python/pgshovel/replication/validation/transactions.py#L141-L143).

Unfortunately, there is no guarantee that the first message is a `BeginOperation`. Most likely[1] this is due to the Kafka topic containing the messages being pruned of old messages, and now the head of the log falls in the middle of a transaction.  And if the first message isn't `BeginOperation`, prior to this diff validation would fail and this code would crash, despite the fact that if the stream just seeked a few more messages, it would have found a `BeginOperation` and could have continued correctly.

[1] It's also possible to call `KafkaStream.consume()` with a `state.stream_state.transaction_state` value that is incompatible with the next message from the `state`'s offset in the mutations stream - i.e. `transaction_state` is `InTransaction` and a `BeginOperation` is the first message seen - though *theoretically* this shouldn't be possible because a replication target instances should be setting `state` offset at the same time as the `transaction_state`. If this does happen, the code will see a `InvalidEventError` on line 72 and either give up trying to find a `BeginOperation` or find one, yield it out, and then will immediately fail the `valid_state` call since the transaction state machine isn't valid and the sequence number of the next message (the found `BeginOperation` isn't 1 more than the `state`'s sequence number).

This may be confusing to you so let me know if this makes sense.